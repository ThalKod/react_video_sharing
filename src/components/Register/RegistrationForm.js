import React from "react";
import { Link } from "react-router-dom";
import axios from "axios";
import { connect } from "react-redux";
import { startSignupUser, startSigninUser } from "../../actions/auth";

class RegistrationForm extends React.Component{

  state = {
    email: "",
    password: "",
    username: "",
    passwordVerification: "",
    errorPass: false,
    errorEmail: false,
    errorUName: false,
  };

  handleSubmit = (e) => {
    e.preventDefault();

    const { signup, signupUser, submit, signinUser } = this.props;
    const { email, password, passwordVerification } = this.state;

    if(signup){
      if(passwordVerification !== password) return this.setState({ errorPass: true});

      return signupUser({ email, password }, (res) => {
        if(!res.error) return submit();
        return res.error;
      })
    }

    return signinUser({ email, password }, (res) => {
      if(!res.error) return submit();
      return res.error;
    });
  };

  // checking if email is already registered
  checkEmail = () => {
    const { email } = this.state;

    axios.post("/api/v0/check/email", { email })
        .then(res => {
          if(!res.data.error){
            if(!res.data.valid) return this.setState({ errorEmail: true });
          }
          return res.data.error;
        })
        .catch(err => console.log(err));
  };

  // checking if username is already registered
  checkUserName = () => {
    const { username } = this.state;
    axios.post("/api/v0/check/username", { username })
        .then(res => {
          if(!res.data.error){
            if(!res.data.valid) return this.setState({ errorUName: true });
          }
          return res.data.error;
        })
        .catch(err => console.log(err));
  };

  passwordMatch = () => {
    const { password, passwordVerification } = this.state;
    if(passwordVerification !== password) return this.setState({ errorPass: true});
    return false;
  };

  render(){
    const { signup } = this.props;
    const { errorEmail, errorPass, errorUName, passwordVerification, password, email, username } = this.state;
    return (
        <div className="login-window">
          {signup ?
              <div className="l-head">
                Sign Up for Free
              </div> :
              <div className="l-head">
                Log in to your Circle Account
              </div>
          }

          <div className="l-form">
            <form>
              <div className="form-group">
                <label  htmlFor="exampleInputEmail1">Email</label>
                { signup && errorEmail && <p style={{ color: "red"}}>Already registered with that email, please try signin !</p>}
                <input
                    value={email}
                    onChange={e => this.setState({ email: e.target.value, errorEmail: false })}
                    onBlur={this.checkEmail}
                    type="email"
                    className="form-control"
                    id="exampleInputEmail1"
                    placeholder="sample@gmail.com"
                />
              </div>
              { signup && errorUName && <p style={{ color: "red"}}>Already registered with that username, please try signin !</p>}
              <div className="form-group">
                <label>Username</label>
                <input
                    value={username}
                    onChange={e => this.setState({ username : e.target.value, errorUName: false })}
                    onBlur={this.checkUserName}
                    type="text"
                    className="form-control"
                    placeholder="username"
                />
              </div>
              <div className="form-group">
                <label htmlFor="exampleInputPassword1">Password</label>
                <input
                    value={password}
                    onChange={e => this.setState({ password : e.target.value, errorPass: false })}
                    type="password"
                    className="form-control"
                    id="exampleInputPassword1"
                    placeholder="**********"
                />
              </div>
              {signup && (
                    <div className="form-group">
                      <label htmlFor="exampleInputPassword2">Re-type Password</label>
                      <input
                          value={passwordVerification}
                          onChange={e => this.setState({ passwordVerification : e.target.value })}
                          onBlur={this.passwordMatch}
                          onFocus={() => this.setState({errorPass: false})}
                          type="password"
                          className="form-control"
                          id="exampleInputPassword2"
                          placeholder="**********"
                      />
                    </div>
              )}
              {errorPass && <p style={{ color: "red" }}>Password don't match !</p>}
              <div className="checkbox">
                <label>
                  <label className="checkbox">
                    <input type="checkbox" name="#"/>
                      <span className="arrow"/>
                  </label> <span>Remember me on this computer</span>
                  <span className="text2">(not recommended on public or shared computers)</span>
                </label>
              </div>
              <div className="row">
                {signup ?
                    <div>
                      <div className="col-lg-7">
                        <button
                            onClick={this.handleSubmit}
                            disabled={ errorPass || errorEmail || errorUName || !username || !email || !password || !passwordVerification}
                            type="submit"
                            className="btn btn-cv1"
                        >Sign Up</button>
                      </div>
                      <div className="col-lg-1 ortext">or</div>
                      <div className="col-lg-4 signuptext">
                        <Link to="/signin">Login</Link>
                      </div>
                    </div>   :
                    <div>
                      <div className="col-lg-7">
                        <button
                            type="submit"
                            onClick={this.handleSubmit}
                            disabled={ !email || !password }
                            className="btn btn-cv1">Login</button>
                      </div>
                      <div className="col-lg-1 ortext">or</div>
                      <div className="col-lg-4 signuptext">
                        <Link to="/signup">Sign Up</Link>
                      </div>
                    </div>
                }
              </div>
              <div className="row">
                <div className="col-lg-12 forgottext">
                  { signup ?
                      <a href="/">By clicking "Sign Up" I agree to circle's Terms of Service.</a> :
                      <a href="/">Forgot Username or Password?</a>
                  }
                </div>
              </div>
            </form>
          </div>
        </div>
    )
  }
}

const mapDispatchToProps = (dispatch) =>({
  signupUser: (user, callback) => dispatch(startSignupUser(user, callback)),
  signinUser: (user, callback) => dispatch(startSigninUser(user, callback)),
});

export default connect(null , mapDispatchToProps)(RegistrationForm);
