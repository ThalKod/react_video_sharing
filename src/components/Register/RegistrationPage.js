import React from "react";

import RegistrationForm from "./RegistrationForm";
import bgLogin from "../../assets/images/login.jpg";

export default function RegistrationPage({ signup, history}){
    return (
        <div>
          <div className="container-fluid bg-image">
            <div className="row">
              <div className="login-wraper">
                <img src={bgLogin} alt="bg-login"/>
                <div className="banner-text">
                  <div className="line"/>
                  <div className="b-text">
                    Watch <span className="color-active">and<br /> upload</span> <span className="color-b1">v</span><span className="color-b2">i</span><span className="color-b3">de</span><span className="color-active">os</span> for free.
                  </div>
                </div>
                <RegistrationForm signup={signup} submit={() => history.push("/")} />
              </div>
            </div>
          </div>
        </div>
    )
}
