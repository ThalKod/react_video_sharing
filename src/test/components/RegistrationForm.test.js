import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";

import { RegistrationForm } from "../../components/Register/RegistrationForm";

describe("RegistrationForm component", () => {

  describe("when login", () => {
    it("should render the correct form", () => {
      const wrapper = shallow(<RegistrationForm/>);
      expect(toJson(wrapper)).toMatchSnapshot();
      expect(
          wrapper.containsMatchingElement(<button>Login</button>)
      ).toBe(true);
    });

    it("should set email on input change", () => {
      const value = "test@mail.com";
      const wrapper = shallow(<RegistrationForm/>);
      wrapper.find("input").at(0).simulate("change", {
        target: { value }
      });
      expect(wrapper.state("email")).toBe(value);
    });

    it("should set password on input change", () => {
      const value = "password";
      const wrapper = shallow(<RegistrationForm/>);
      wrapper.find("input").at(1).simulate("change", {
        target: { value }
      });
      expect(wrapper.state("password")).toBe(value);
    });

    it("Login button should be disable if email and password not set", () => {
      const wrapper = shallow(<RegistrationForm />);
      const loginButton = wrapper.find("button").first();
      expect(
          loginButton.props().disabled
      ).toBe(true);
    });

    it("Login button should be enable if email and password completed", () => {
      const wrapper = shallow(<RegistrationForm />);
      wrapper.setState({ email: "test@mail.com", password: "password" });
      const loginButton = wrapper.find("button").first();
      expect(
          loginButton.props().disabled
      ).toBe(false);
    });

    it("should submit valid form", () => {
      const signinUser = jest.fn();
      const user = {email: "test@mail.com", password: "password"};
      const wrapper = shallow(<RegistrationForm signinUser={signinUser} />);

      wrapper.setState({...user});
      wrapper.find("button").first().simulate("click", {
        preventDefault: () => {}
      });
      expect(signinUser).toHaveBeenCalled();
    });

  });

  describe("when signup", () => {
    it("should render the correct form if we want to signup", () => {
      const wrapper = shallow(<RegistrationForm signup/>);
      expect(toJson(wrapper)).toMatchSnapshot();
      expect(
          wrapper.containsMatchingElement(<button>Sign Up</button>)
      ).toBe(true);
    });

    it("should set email on input change", () => {
      const value = "test@mail.com";
      const wrapper = shallow(<RegistrationForm signup/>);
      wrapper.find("input").at(0).simulate("change", {
        target: { value }
      });
      expect(wrapper.state("email")).toBe(value);
    });

    it("should set username on input change", () => {
      const value = "username";
      const wrapper = shallow(<RegistrationForm signup/>);
      wrapper.find("input").at(1).simulate("change", {
        target: { value }
      });
      expect(wrapper.state("username")).toBe(value);
    });

    it("should set password on input change", () => {
      const value = "password";
      const wrapper = shallow(<RegistrationForm signup/>);
      wrapper.find("input").at(2).simulate("change", {
        target: { value }
      });
      expect(wrapper.state("password")).toBe(value);
    });

    it("should set passwordverification on input change", () => {
      const value = "passwordveerif";
      const wrapper = shallow(<RegistrationForm signup/>);
      wrapper.find("input").at(3).simulate("change", {
        target: { value }
      });
      expect(wrapper.state("passwordVerification")).toBe(value);
    });

    it("signup button should be disable if form not complete", () => {
      const wrapper = shallow(<RegistrationForm signup />);
      const loginButton = wrapper.find("button").first();
      expect(
          loginButton.props().disabled
      ).toBe(true);
    });

    it("signup button should be disable if there's an error", () => {
      const wrapper = shallow(<RegistrationForm signup />);
      wrapper.setState({ errorEmail: true });
      const loginButton = wrapper.find("button").first();
      expect(
          loginButton.props().disabled
      ).toBe(true);
    });

    it("signup button should be enable if all field complete and no error", () => {
      const wrapper = shallow(<RegistrationForm signup />);
      wrapper.setState({ email: "test@mail.co", password: "password", passwordVerification: "password", username: "password" });
      const loginButton = wrapper.find("button").first();
      expect(
          loginButton.props().disabled
      ).toBe(false);
    });
  });
});
