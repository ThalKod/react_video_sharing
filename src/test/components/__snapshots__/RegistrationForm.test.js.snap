// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegistrationForm component when login should render the correct form 1`] = `
<div
  className="login-window"
>
  <div
    className="l-head"
  >
    Log in to your Circle Account
  </div>
  <div
    className="l-form"
  >
    <form>
      <div
        className="form-group"
      >
        <label
          htmlFor="exampleInputEmail1"
        >
          Email
        </label>
        <input
          className="form-control"
          id="exampleInputEmail1"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="sample@gmail.com"
          type="email"
          value=""
        />
      </div>
      <div
        className="form-group"
      >
        <label
          htmlFor="exampleInputPassword1"
        >
          Password
        </label>
        <input
          className="form-control"
          id="exampleInputPassword1"
          onChange={[Function]}
          placeholder="**********"
          type="password"
          value=""
        />
      </div>
      <div
        className="checkbox"
      >
        <label>
          <label
            className="checkbox"
          >
            <input
              name="#"
              type="checkbox"
            />
            <span
              className="arrow"
            />
          </label>
           
          <span>
            Remember me on this computer
          </span>
          <span
            className="text2"
          >
            (not recommended on public or shared computers)
          </span>
        </label>
      </div>
      <div
        className="row"
      >
        <div>
          <div
            className="col-lg-7"
          >
            <button
              className="btn btn-cv1"
              disabled={true}
              onClick={[Function]}
              type="submit"
            >
              Login
            </button>
          </div>
          <div
            className="col-lg-1 ortext"
          >
            or
          </div>
          <div
            className="col-lg-4 signuptext"
          >
            <Link
              replace={false}
              to="/signup"
            >
              Sign Up
            </Link>
          </div>
        </div>
      </div>
      <div
        className="row"
      >
        <div
          className="col-lg-12 forgottext"
        >
          <a
            href="/"
          >
            Forgot Username or Password?
          </a>
        </div>
      </div>
    </form>
  </div>
</div>
`;

exports[`RegistrationForm component when signup should render the correct form if we want to signup 1`] = `
<div
  className="login-window"
>
  <div
    className="l-head"
  >
    Sign Up for Free
  </div>
  <div
    className="l-form"
  >
    <form>
      <div
        className="form-group"
      >
        <label
          htmlFor="exampleInputEmail1"
        >
          Email
        </label>
        <input
          className="form-control"
          id="exampleInputEmail1"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="sample@gmail.com"
          type="email"
          value=""
        />
      </div>
      <div
        className="form-group"
      >
        <label>
          Username
        </label>
        <input
          className="form-control"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="username"
          type="text"
          value=""
        />
      </div>
      <div
        className="form-group"
      >
        <label
          htmlFor="exampleInputPassword1"
        >
          Password
        </label>
        <input
          className="form-control"
          id="exampleInputPassword1"
          onChange={[Function]}
          placeholder="**********"
          type="password"
          value=""
        />
      </div>
      <div
        className="form-group"
      >
        <label
          htmlFor="exampleInputPassword2"
        >
          Re-type Password
        </label>
        <input
          className="form-control"
          id="exampleInputPassword2"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="**********"
          type="password"
          value=""
        />
      </div>
      <div
        className="checkbox"
      >
        <label>
          <label
            className="checkbox"
          >
            <input
              name="#"
              type="checkbox"
            />
            <span
              className="arrow"
            />
          </label>
           
          <span>
            Remember me on this computer
          </span>
          <span
            className="text2"
          >
            (not recommended on public or shared computers)
          </span>
        </label>
      </div>
      <div
        className="row"
      >
        <div>
          <div
            className="col-lg-7"
          >
            <button
              className="btn btn-cv1"
              disabled={true}
              onClick={[Function]}
              type="submit"
            >
              Sign Up
            </button>
          </div>
          <div
            className="col-lg-1 ortext"
          >
            or
          </div>
          <div
            className="col-lg-4 signuptext"
          >
            <Link
              replace={false}
              to="/signin"
            >
              Login
            </Link>
          </div>
        </div>
      </div>
      <div
        className="row"
      >
        <div
          className="col-lg-12 forgottext"
        >
          <a
            href="/"
          >
            By clicking "Sign Up" I agree to circle's Terms of Service.
          </a>
        </div>
      </div>
    </form>
  </div>
</div>
`;
